### PR TITLE
feat: typescript

### PR DESCRIPTION
**Pull request by lodmfjord**
Added typescript

### CODE REVIEW

**Comment by lodmfjord** [comment]
Oh god! I didn't notice you had already created a typescript branch. Sorry.

**Comment by PandaSekh** [comment]
No worries, and thank you! I'm a typescript newbie so I'm sure your version is better. I already noticed you managed the custom event, while I had problems with that. I'm gonna review this soon.

**Comment by lodmfjord** [comment]
the tests are not running correctly haven't had time to look into it yet :)

**Comment by lodmfjord** [comment]
The tests need to have tsconfig react-jsx but the compile process need it to be preserve.

Will have to run tests with different tsconfig. Will look into it later today.

**Comment by PandaSekh** [comment]
Changes: 
- Fixed some imports that were still ".js"
- Fixed a small typo where the decrease dispatch action instead was increasing the counter


Probably the build won't work because react scripts overwrites the tsconfig.json file

**Comment by PandaSekh** [comment]
Same as other branch.
React scripts change the tsconfig:
```
The following changes are being made to your tsconfig.json file:
  - compilerOptions.jsx must be react-jsx (to support the new JSX transform in React 17)
```

It runs the tests, they pass, then tries to build and this error occur:
```
rpt2: config error TS6046: Argument for '--jsx' option must be: 'preserve', 'react-native', 'react'.
(rpt2 plugin) Error: /home/travis/build/PandaSekh/lepre/src/index.ts(1,19): semantic error TS6142: Module './Emoji' was resolved to '/home/travis/build/PandaSekh/lepre/src/Emoji.tsx', but '--jsx' is not set.
```

**Comment by lodmfjord** [comment]
Yeah jsx should be preserve to work.

**Comment by lodmfjord** [comment]
The jsx thing should have fixed. Pull again.

**Comment by PandaSekh** [comment]
I'm using preserve as you suggested, react scripts just overwrites it

**Comment by lodmfjord** [comment]
Yes, we need to find a solution for that :)

**Comment by lodmfjord** [comment]
ok pushing a fix

**Comment by lodmfjord** [comment]
OK two different package.json for test and building which are extended from base tsconfig

**Comment by PandaSekh** [comment]
Great job! Only problem now is bundle size, as the files now exceeds the 1Kb mark I wanted this library to stay below.  

Do you maybe know how this is possible and how we could possibly fix this? Is this because of the typescript compiler?

**Comment by lodmfjord** [comment]
Wait I will look into it.

**Comment by lodmfjord** [comment]
OK,  this is weird:
My old builds are:
index.js 2.4kb
index.modern.js 2.3kb

New builds:
index.js 2.6kb
index.modern.js 2.5kb

What sizes do you get?

I notice most files are same size or equal in ts but two files are bigger, going to check out why.

**Comment by PandaSekh** [comment]
I thought it was weird too, but I think you are looking at the actual file size, right? I got my numbers from the console output after the build and those file are gzipped. If I check the actual files, I get the same sizes as you. Still, we both get about 200 bytes extra just because we added typescript, that's strange.

The Github action to check file size changes reports this:  

**Size Change:** +218 B (+11%) ⚠️

**Total Size:** 2.12 kB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/index.js` | 1.06 kB | +110 B (+12%) | ⚠️ |
| `dist/index.modern.js` | 1.05 kB | +108 B (+11%) | ⚠️ |


On my machine if I build I get this: 
Before (master):  
926 B: index.js.gz
912 B: index.modern.js.gz

After (ts support): 
1035 B: index.js.gz
1021 B: index.modern.js.gz

**Comment by PandaSekh** [comment]
This quick fix changes the tsconfig.json includes to `"include": ["src/index.ts", "src/types.ts"]`, because it was including in the output directory tests too

**Comment by lodmfjord** [comment]
I am viewing the difference with sourcemap and seeing what I can do :)

**Comment by lodmfjord** [comment]
reading a little about your builder haven't tried it before

**Comment by PandaSekh** [comment]
You mean microbuilder?

Btw I noticed that my typescript branch even reduced file size by about 5bytes. My typescript code is pretty bad and I had problems with useEmojis and the custom event, but maybe we can look at that to see if there's any obvious difference. I'll do that later or tomorrow

**Comment by lodmfjord** [comment]
Most files are reduced by small amount, but there are two files that have increased their filesize because of a change I did. I could revert that change.

**Comment by lodmfjord** [comment]
Will try to change that back later.

**Comment by lodmfjord** [comment]
Going to change the tests to ts first

**Comment by lodmfjord** [comment]
OK @PandaSekh check this out :)

File: src/EmojiBlock.tsx [hunk 1]
```diff
@@ -21,16 +21,14 @@ export default function EmojiBlock({
     }
   }, [state]);
 
-  const mappedReactions = state
-    .sort((a, b) => (a.emoji < b.emoji ? 1 : -1))
```
Comment (PandaSekh) [inline src/EmojiBlock.tsx:17]: Unfortunately the sort is necessary, otherwise on each click the emojis would change position.

Comment (lodmfjord) [reply src/EmojiBlock.tsx:17]: sorry! Wasn't gonna delete that was just testing.

**Comment by lodmfjord** [comment]
Sort back in :)

**Comment by PandaSekh** [comment]
Thanks, I just made the same commit 😂 gonna revert.  

Unfortunately that sort is the only thing I can't test, in fact the coverage is at 98% just because of it.

Thanks for the great work, we should be back below 1Kb!

**Comment by lodmfjord** [comment]
Haha! This was fun.

Good luck with your project, I will be watching.

**Comment by PandaSekh** [comment]
Thank you!

**Pull request merged**